gapis/api/vulkan/graph_visualization.go

```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
 	return subCommandHierarchyNames
 }
 
-func getCommandBuffer(command api.Cmd) string {
+func getMaxCommonPrefix(label1 *api.Label, label2 *api.Label) int {
+	size := len(label1.LevelsID)
+	if len(label2.LevelsID) < size {
+		size = len(label2.LevelsID)
+	}
+	for i := 0; i < size; i++ {
+		if label1.LevelsName[i] != label2.LevelsName[i] || label1.LevelsID[i] != label2.LevelsID[i] {
+			return i
+		}
+	}
+	return size
+}
+
+func addDebugMarker(builder *labelForVulkanCommands, from, to int) {
+	level := builder.labelsInsideDebugMarkers[len(builder.labelsInsideDebugMarkers)-1].GetSize() - 1
+	builder.numberOfDebugMarker++
+	for i := from; i <= to; i++ {
+		builder.labelsInsideDebugMarkers[i].Insert(level, VK_CMD_DEBUG_MARKER, builder.numberOfDebugMarker)
+	}
+}
+
+func checkDebugMarkers(builder *labelForVulkanCommands, currentLabel *api.Label) {
+	commandName := currentLabel.GetCommandName()
+	positionOfLastDebugMarkerBegin := 0
+	labelOfLastDebugMarkerBegin := &api.Label{}
+	if len(builder.positionOfDebugMarkersBegin) > 0 {
+		positionOfLastDebugMarkerBegin = builder.positionOfDebugMarkersBegin[len(builder.positionOfDebugMarkersBegin)-1]
+		labelOfLastDebugMarkerBegin = builder.labelsInsideDebugMarkers[positionOfLastDebugMarkerBegin]
+	}
+
+	if commandName == VK_CMD_DEBUG_MARKER_BEGIN {
+		if len(builder.positionOfDebugMarkersBegin) > 0 {
+			if labelOfLastDebugMarkerBegin.GetSize() != currentLabel.GetSize() {
+				builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:0]
+			}
+		}
+		builder.labelsInsideDebugMarkers = append(builder.labelsInsideDebugMarkers, currentLabel)
+		builder.positionOfDebugMarkersBegin = append(builder.positionOfDebugMarkersBegin, len(builder.labelsInsideDebugMarkers)-1)
+
+	} else if commandName == VK_CMD_DEBUG_MARKER_END {
+
+		if len(builder.positionOfDebugMarkersBegin) > 0 {
+			if labelOfLastDebugMarkerBegin.GetSize() != currentLabel.GetSize() {
+				builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:0]
+
+			} else if getMaxCommonPrefix(labelOfLastDebugMarkerBegin, currentLabel) == currentLabel.GetSize()-1 {
+				builder.labelsInsideDebugMarkers = append(builder.labelsInsideDebugMarkers, currentLabel)
+				builder.positionOfDebugMarkersBegin = append(builder.positionOfDebugMarkersBegin, len(builder.labelsInsideDebugMarkers)-1)
+				addDebugMarker(builder, positionOfLastDebugMarkerBegin, len(builder.labelsInsideDebugMarkers)-1)
+				builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:len(builder.positionOfDebugMarkersBegin)-1]
+			}
+		}
+
+	} else if len(builder.positionOfDebugMarkersBegin) > 0 {
+		if labelOfLastDebugMarkerBegin.GetSize() != currentLabel.GetSize() {
+			builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:0]
```

Comment:
We have a ton of code in vulkan.go to handle nesting of debug markers and similar, (as used for displaying groups in the UI) It handles some of these cases (bad nestings etc).

Is there a chance we can leverage that? 
It may be a follow-up, but it may save some time in handling some of these cases.

Reply:
Now I am adding debug markers as long as it does not destroy any level of hierarchy defined in commands. So, if exists some command between debug_marker_begin and debug_marker_end which has smaller Label size , adding the debug marker here will destroy the levels of hierarchy and for this case I just ignore the debug marker. It means I am preserving Labels with no-decreasing size. I have tested with several captures and it is working very nice. In addition to this, I am analizing the possibility to include code from vulkan.go to addres this problem as well, but if it is the case , I will add this in the last commit.